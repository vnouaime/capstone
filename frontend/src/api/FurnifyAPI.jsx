import axios from "axios"
import { BASE_URL } from "../config";

// Own database API, handling users, carts, and orders. 

class FurnifyAPI {

  // the token for the API will be stored here.
  static token;

  static async request(endpoint, data = {}, method = "get") {
    console.debug("API Call:", endpoint, data, method);

    const url = `${BASE_URL}/${endpoint}`;
    const headers = { Authorization: `Bearer ${FurnifyAPI.token}` };
    const params = (method === "get")
        ? data
        : {};

    try {

      return (await axios({ url, method, data, params, headers })).data;
    } catch (err) {
      console.error("API Error:", err.response);
      let message = err.response.data.error.message;

      throw Array.isArray(message) ? message : [message];
    }
  }

  /** Users *******************************************************************************/

  /** Get the current user. */
  static async getCurrentUser(username) {
    let res = await this.request(`users/${username}`);
    return res.user;
  }

  /** Get token for login from username, password. */
  static async login(data) {
    let res = await this.request(`auth/token`, data, "post");
    
    return res.token;
  }

  /** Signup for site. */
  static async signup(data) {
    let res = await this.request(`auth/register`, data, "post");
    
    return res.token;
  }

  /** Save user profile page. */
  static async saveProfile(username, data) {
    let res = await this.request(`users/${username}`, data, "patch");
    return res.user;
  }

  /** Carts *******************************************************************************/

  /** 
   * Handles, get, post, and patch requests.
   * 
   * GET: Gets cart based on current user.
   * POST: Creates new cart when user signs up.
   * PATCH: Adds or removes product sku from products and updates cart total. 
  */
  static async cart(username, data, action) {
    let res = await this.request(`cart/${username}`, data, action)
    return res.cart
  }
  
  /** Orders *******************************************************************************/

  /** 
   * Handles, get, and post requests.
   * 
   * GET: Gets orders based on current user.
   * POST: Creates new order. 
  */
  static async order(username, data, action) {
    let res = await this.request(`orders/${username}`, data, action)

    if (action == "post") {
      return res.order
    }

    if (action == "get") {
      return res.orders
    }
    
  }
}

export default FurnifyAPI;