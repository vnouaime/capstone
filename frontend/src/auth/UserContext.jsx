import React from "react";

/** Provides current user object to be used throughout browser. */

const UserContext = React.createContext();

export default UserContext;