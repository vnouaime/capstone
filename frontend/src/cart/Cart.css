.cart-header {
    margin-top: 30px;
}

.cart-checkout-header {
    font-size: 35px;
    margin: 15px;
}

.cart-empty {
    font-size: xx-large;
    margin-top: 100px;
}