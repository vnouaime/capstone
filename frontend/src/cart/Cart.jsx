import React, { useState, useContext, useEffect } from "react";
import { Container, Row, Col } from "react-bootstrap"

import UserContext from "../auth/UserContext";
import FurnifyAPI from "../api/FurnifyAPI";
import ExternalAPI from "../api/ExternalAPI"
import CartProduct from "./CartProduct";
import CartTotal from "./CartTotal";
import LoadingSpinner from "../common/LoadingSpinner";
import Alert from "../common/Alert";

import "./Cart.css"

/** Cart.
 * 
 * Renders current user's cart. Shows items and quantity in cart. Can edit quanity per item in cart and displays cart total. Alert message will appear when adding and removing products from cart. 
 * 
 * Props:
 * none
 * 
 * State: 
 * cart, alertMessage, alertType
 * 
 * Routes -> Cart -> CartProduct, CartTotal, LoadingSpinner, Alert
 * Routed as /cart/:username
 */
const Cart = () => {
    const { currentUser } = useContext(UserContext);
    const [cart, setCart] = useState(null);
    const [alertMessage, setAlertMessage] = useState(null)
    const [alertType, setAlertType] = useState(null)
    
    useEffect(() => {
        async function getUserCart() {
            try {
                const username = currentUser.username;
                const cartData = await FurnifyAPI.cart(username, {}, "get");
                let updatedCartData;

                if (cartData.products) {
                    const productsDetails = await Promise.all(cartData.products.map(async productSku => {
                        // Fetch details for each product
                        const productDetails = await ExternalAPI.getProductDetails(productSku);
                        return productDetails;
                    }));
                    
                    // Update cartData with product details. Reduces productDetails to array with quantity of product in cart to avoid duplication. 
                    updatedCartData = { 
                        ...cartData, 
                        products: productsDetails.reduce((acc, product) => {
                            const existingProductIndex = acc.findIndex(p => p.sku === product.sku);
                            if (existingProductIndex !== -1) {
                                acc[existingProductIndex].quantity += 1;
                            } else {
                                acc.push({ ...product, quantity: 1 });
                            }
                            return acc;
                        }, [])  
                    };
    
                } else {
                    updatedCartData = { ...cartData }
                }

                setCart(updatedCartData);
                
            } catch (error) {
                console.error("Error fetching cart:", error);
            }
        }
        getUserCart();
    }, [currentUser]); 

    if (!cart) return <LoadingSpinner />

    /** Adds product to cart and updates state by editing quantity and total */
    const handleAddToCart = async (sku, productPrice) => {
        try {
            const data = { "productSku": sku, "action": "add", productPrice };
            const res = await FurnifyAPI.cart(currentUser.username, data, "patch");

            if (res && res.products) {
                setCart(prevCart => {
                    const updatedProducts = prevCart.products.map(product => {
                        if (product.sku === sku) {
                            return { ...product, quantity: product.quantity + 1 };
                        } else {
                            return product; 
                        }
                    });

                    return {
                        ...prevCart,
                        products: updatedProducts,
                        total: res.total 
                    };
                });

                setAlertMessage("Added to Cart");
                setAlertType("success");
            }

            return res;
        } catch (error) {
            console.error("Error adding to cart:", error);
            setAlertMessage("Error adding product to cart");
            setAlertType("danger");
        }
    };


    /** Removes product from cart and updates state by editing quantity and total */
    const handleRemoveFromCart = async (sku, price) => {
        try {
            const data = { "productSku": sku, "action": "remove", productPrice: price };
            const res = await FurnifyAPI.cart(currentUser.username, data, "patch");

            if (res && res.products) {
                setCart(prevCart => ({
                    ...prevCart,
                    products: prevCart.products.map(product => {
                        if (product.sku === sku) {
                            // If quantity is more than 1, decrease it by 1
                            if (product.quantity > 1) {
                                return { ...product, quantity: product.quantity - 1 };
                            } else {
                                // If quantity is 1, remove the product from the cart
                                return null;
                            }
                        } else {
                            return product;
                        }
                    }).filter(Boolean), // Filter out null
                    total: res.total
                }));

                setAlertMessage("Product removed from cart");
                setAlertType("success"); 
            }

            return res;
        } catch (error) {
            console.error("Error removing from cart:", error);
            setAlertMessage("Error removing product from cart"); 
            setAlertType("danger");
        }
    };

    return (
        <div>
            <Container fluid>
                {alertMessage && <Alert type={alertType} messages={[alertMessage]} />}
                <Row className="mb-5">
                    <h1 className="cart-header text-center">My Cart</h1>
                    <hr />
                </Row> 
                {cart && cart.products && cart.products.length > 0 ? (
                    <Row>
                        <Col md={9}>
                            {cart.products.map((product) => (
                                <div key={product.sku}> 
                                    <CartProduct 
                                        sku={product.sku} 
                                        name={product.name}
                                        averageRatingValue={product.customer_reviews.average_rating_value}
                                        totalCustomerReviews={product.customer_reviews.reviews.length}
                                        price={product.unitPrice.customerPrice}
                                        quantity={product.quantity}
                                        onAdd={handleAddToCart}
                                        onRemove={handleRemoveFromCart} 
                                    />
                                </div>
                            ))}
                        </Col>
                        <Col>
                            <CartTotal total={cart.total} buttonName="Checkout" buttonRoute="/checkout"/> 
                        </Col>
                    </Row>
                ) : (
                    <Row>
                        <div className="cart-empty text-center">
                            No items in cart
                        </div>
                    </Row>
                )}
            </Container>
        </div>
    );
};

export default Cart;
