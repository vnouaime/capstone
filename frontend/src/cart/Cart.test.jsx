import React from "react"
import { beforeAll, afterEach, expect, vi } from "vitest"
import { render, fireEvent } from "vitest-react"
import UserContext from "../auth/UserContext"
import Cart from "./Cart"
import { MemoryRouter } from "react-router-dom";
import FurnifyAPI from "../api/FurnifyAPI"
import FurnitureAPI from "../api/ExternalAPI"

const currentUser = {
    username: "testuser"
}

FurnifyAPI.cart = vi.fn()
FurnitureAPI.getProductDetails = vi.fn()

afterEach(() => (
    vi.clearAllMocks()
))

test("smoke test", () => {
    const mockCartData = {
        username: currentUser.username,
        products: [],
        total: 0,
    };

    FurnifyAPI.cart.mockResolvedValue(mockCartData);

    render(
        <UserContext.Provider value={{ currentUser }}>
            <Cart />
        </UserContext.Provider>
    )
})

test("snapshot", () => {
    const mockCartData = {
        username: currentUser.username,
        products: [],
        total: 0,
    };

    FurnifyAPI.cart.mockResolvedValue(mockCartData);

    const { asFragment } = render(
        <UserContext.Provider value={{ currentUser }}>
            <Cart />
        </UserContext.Provider>
    )
    expect(asFragment()).toMatchSnapshot()
})

test('renders cart products when data is fetched', async () => {
    const mockCartData = {
        username: currentUser.username,
        products: [
          { sku: '123', name: 'Product 1', unitPrice: { customerPrice: 10 }, quantity: 2 },
          { sku: '456', name: 'Product 2', unitPrice: { customerPrice: 20 }, quantity: 1 },
        ],
        total: 40,
    };
  
    const mockProductDetails = {
        sku: '1',
        name: 'Product 1',
        customer_reviews: {
          average_rating_value: 4.5,
          reviews: [{}, {}],
        },
        unitPrice: { customerPrice: 50 },
    };

    FurnifyAPI.cart.mockResolvedValue(mockCartData);
    FurnitureAPI.getProductDetails.mockResolvedValue(mockProductDetails)
  
    const { findByText } = render(
        <MemoryRouter>
            <UserContext.Provider value={{ currentUser }}>
                <Cart />
            </UserContext.Provider>
        </MemoryRouter>
        
    );
  
    const cart = await findByText("My Cart")
    const product1 = await findByText("Product 1")

    expect(cart).toBeInTheDocument()
    expect(product1).toBeInTheDocument()
})

test('Displays no items in cart if there are none', async () => {
    const mockCartData = {
        username: currentUser.username,
        products: [],
        total: 0,
    };
  
    FurnifyAPI.cart.mockResolvedValue(mockCartData);
  
    const { findByText } = render(
        <MemoryRouter>
            <UserContext.Provider value={{ currentUser }}>
                <Cart />
            </UserContext.Provider>
        </MemoryRouter>
        
    );
  
    const cart = await findByText("My Cart")
    const emptyCart = await findByText("No items in cart")

    expect(cart).toBeInTheDocument()
    expect(emptyCart).toBeInTheDocument()
})
