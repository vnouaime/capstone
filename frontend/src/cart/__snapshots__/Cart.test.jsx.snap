// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`snapshot 1`] = `
<DocumentFragment>
  <div
    class="d-flex justify-content-center align-items-center container"
    style="height: 100vh;"
  >
    <div
      class="loadingSpinner spinner-border"
      role="status"
    >
      <span
        class="visually-hidden"
      >
        Loading...
      </span>
    </div>
  </div>
</DocumentFragment>
`;
